zokrates_core/src/static_analysis/zir_propagation.rs

```diff
@@ -0,0 +1,1414 @@
+                    i => UExpressionInner::Select(e, box i.annotate(bitwidth)),
+                }
+            }
+            UExpressionInner::Add(box e1, box e2) => {
+                let e1 = self.fold_uint_expression(e1);
+                let e2 = self.fold_uint_expression(e2);
+
+                match (e1.into_inner(), e2.into_inner()) {
+                    (UExpressionInner::Value(0), e) | (e, UExpressionInner::Value(0)) => e,
+                    (UExpressionInner::Value(n1), UExpressionInner::Value(n2)) => {
+                        UExpressionInner::Value((n1 + n2) % 2_u128.pow(bitwidth.to_usize() as u32))
+                    }
+                    (e1, e2) => {
+                        UExpressionInner::Add(box e1.annotate(bitwidth), box e2.annotate(bitwidth))
+                    }
+                }
+            }
+            UExpressionInner::Sub(box e1, box e2) => {
+                let e1 = self.fold_uint_expression(e1);
+                let e2 = self.fold_uint_expression(e2);
+
+                match (e1.into_inner(), e2.into_inner()) {
+                    (e, UExpressionInner::Value(0)) => e,
+                    (UExpressionInner::Value(n1), UExpressionInner::Value(n2)) => {
+                        UExpressionInner::Value(
+                            n1.wrapping_sub(n2) % 2_u128.pow(bitwidth.to_usize() as u32),
+                        )
+                    }
+                    (e1, e2) => {
+                        UExpressionInner::Sub(box e1.annotate(bitwidth), box e2.annotate(bitwidth))
+                    }
+                }
+            }
+            UExpressionInner::Mult(box e1, box e2) => {
+                let e1 = self.fold_uint_expression(e1);
+                let e2 = self.fold_uint_expression(e2);
+
+                match (e1.into_inner(), e2.into_inner()) {
+                    (_, UExpressionInner::Value(0)) | (UExpressionInner::Value(0), _) => {
+                        UExpressionInner::Value(0)
+                    }
+                    (e, UExpressionInner::Value(1)) | (UExpressionInner::Value(1), e) => e,
+                    (UExpressionInner::Value(n1), UExpressionInner::Value(n2)) => {
+                        UExpressionInner::Value((n1 * n2) % 2_u128.pow(bitwidth.to_usize() as u32))
+                    }
+                    (e1, e2) => {
+                        UExpressionInner::Mult(box e1.annotate(bitwidth), box e2.annotate(bitwidth))
+                    }
+                }
+            }
+            UExpressionInner::Div(box e1, box e2) => {
+                let e1 = self.fold_uint_expression(e1);
+                let e2 = self.fold_uint_expression(e2);
+
+                match (e1.into_inner(), e2.into_inner()) {
+                    (UExpressionInner::Value(n1), UExpressionInner::Value(n2)) => {
+                        UExpressionInner::Value((n1 / n2) % 2_u128.pow(bitwidth.to_usize() as u32))
+                    }
+                    (e1, e2) => {
```

Comment:
`a / 1 = a`

Reply:
We cannot add `0 / a = 0` because a could be 0, so we keep this as a runtime error.

Reply:
we could match for `0 / a = 0 if a != 0` no?

Reply:
Well, unless `a` is a constant you cannot tell if `a == 0`. So yeah why not but this is quite niche.